x/ibc/23-commitment/types/merkle.go

```diff
@@ -128,25 +137,117 @@ func (MerkleProof) GetCommitmentType() exported.Type {
 }
 
 // VerifyMembership verifies the membership pf a merkle proof against the given root, path, and value.
-func (proof MerkleProof) VerifyMembership(root exported.Root, path exported.Path, value []byte) error {
+func (proof MerkleProof) VerifyMembership(specs []string, root exported.Root, path exported.Path, value []byte) error {
 	if proof.IsEmpty() || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() || len(value) == 0 {
 		return sdkerrors.Wrap(ErrInvalidMerkleProof, "empty params or proof")
 	}
 
+	if len(specs) != len(proof.Proof.Ops) {
+		return sdkerrors.Wrapf(ErrInvalidMerkleProof,
+			"length of specs: %d not equal to length of proof: %d",
+			len(specs), len(proof.Proof.Ops))
+	}
+
+	for i, op := range proof.Proof.Ops {
+		if op.Type != specs[i] {
+			return sdkerrors.Wrapf(ErrInvalidMerkleProof,
+				"proof does not match expected format at position %d, expected: %s, got: %s",
+				i, specs[i], op.Type)
+		}
+	}
+
 	runtime := rootmulti.DefaultProofRuntime()
 	return runtime.VerifyValue(proof.Proof, root.GetHash(), path.String(), value)
 }
 
 // VerifyNonMembership verifies the absence of a merkle proof against the given root and path.
-func (proof MerkleProof) VerifyNonMembership(root exported.Root, path exported.Path) error {
+func (proof MerkleProof) VerifyNonMembership(specs []string, root exported.Root, path exported.Path) error {
 	if proof.IsEmpty() || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() {
 		return sdkerrors.Wrap(ErrInvalidMerkleProof, "empty params or proof")
 	}
 
+	if len(specs) != len(proof.Proof.Ops) {
+		return sdkerrors.Wrapf(ErrInvalidMerkleProof,
+			"length of specs: %d not equal to length of proof: %d",
+			len(specs), len(proof.Proof.Ops))
+	}
+
+	for i, op := range proof.Proof.Ops {
+		if op.Type != specs[i] {
+			return sdkerrors.Wrapf(ErrInvalidMerkleProof,
+				"proof does not match expected format at position %d, expected: %s, got: %s",
+				i, specs[i], op.Type)
+		}
+	}
```

Comment:
Nit, move this to a `func validateVerificationParams(proof exported.Proof, root exported.Root, specs []string) error` function:

```suggestion
	if err := validateVerificationParams(proof, root, specs); err != nil {
		return err
	}

	if path == nil || path.IsEmpty() {
		return sdkerrors.Wrap(ErrInvalidMerkleProof, "empty path")
	}
```